### PR TITLE
Implement the RigL pruner for image classification

### PR DESCRIPTION
**Pull request by ohaijen**
Implementation of the RigL pruner. (https://arxiv.org/abs/1911.11134)

Testing: added tests + ran a 100-epoch 90% sparse run, getting comparable top-1 accuracy drop to original paper (mine: 4.3%, theirs: 3.7%).

### CODE REVIEW

File: src/sparseml/pytorch/sparsification/pruning/modifier_pruning_rigl.py [hunk 1]
```diff
@@ -0,0 +1,599 @@
...
+    ):
+        # enable grad buffering
+        torch.cuda.empty_cache()
+        if self._scorer._is_main_proc:
+            self._scorer._enabled_grad_buffering = True
+        # reset masks
+        for mask in self._module_masks._param_masks:
+            mask.data = torch.ones_like(mask)
+        # collect grad
+        self._collect_grad_samples(module, self._grad_sampler)
```
Comment (anmarques) [inline src/sparseml/pytorch/sparsification/pruning/modifier_pruning_rigl.py:441]: If I understand this correctly we're accumulating the gradients between mask updates. What I understood from the paper is that this only done in the steps where the mask is updated (and thus avoid computing "dense" gradients every step).

I think that if we do it on during the mask update we can simplify the implementation since if fits the base class premise more closely

**Review by anmarques — CHANGES_REQUESTED** [review]

File: src/sparseml/pytorch/torchvision/train.py [hunk 2]
```diff
@@ -617,19 +630,38 @@ def log_metrics(tag: str, metrics: utils.MetricLogger, epoch: int, epoch_step: i
                 f"{tag}/{metric_name}", smoothed_value.global_avg, step=step
             )
 
+    recipe_kwargs = {}
+    # TODO: What is the right logic to check if we need a grad sampler here? It seems
+    # that for YOLO the grad sampler is always created, whether or not it's needed. See
+    # https://github.com/neuralmagic/sparseml/blob/b73a173ff89c3bb524dcc0a1f7a16a3109a234a1/src/sparseml/yolov8/trainers.py#L699
+    if True:
```
[the remark this answers is not in the document]
Comment (ohaijen) [reply src/sparseml/pytorch/torchvision/train.py:637]: I put that there to signal that we might want some better condition than just True, and wrote the TODO just above it. I took out the if True, but would you like me to leave the TODO or also remove it?

File: src/sparseml/pytorch/sparsification/pruning/modifier_pruning_rigl.py [hunk 3]
```diff
@@ -0,0 +1,599 @@
...
+    def get_param_score(self, param: Tensor, param_grad: Tensor, param_sparsity: float):
+        """
+        Computed the saliency score for a given parameter.
+
+        :param param: param to be scored
+        :param param_grad: gradient of the parameters
+        :param param_sparsity: the sparsity for a given parameter
+        """
+        # We drop and replace update_fraction of nonzero elements
+        updated_number = (1 - param_sparsity) * self._update_fraction
```
[the remark this answers is not in the document]
Comment (ohaijen) [reply src/sparseml/pytorch/sparsification/pruning/modifier_pruning_rigl.py:567]: It's not a decrement, because sparsity is always constant. Rather, it's the proportion of the mask that can also change during the update. I changed it to mask_update_fraction.

File: src/sparseml/pytorch/sparsification/pruning/modifier_pruning_rigl.py [hunk 4]
```diff
@@ -0,0 +1,599 @@
...
+            mask.data = torch.ones_like(mask)
+        # collect grad
+        self._collect_grad_samples(module, self._grad_sampler)
+        self._pre_step_completed = True
+        # disable grad buffering
+        torch.cuda.empty_cache()
+        if self._scorer._is_main_proc:
+            self._scorer._enabled_grad_buffering = False
+
+        self.scorer._update_fraction = self._get_update_fraction(epoch)
```
[the remark this answers is not in the document]
Comment (ohaijen) [reply src/sparseml/pytorch/sparsification/pruning/modifier_pruning_rigl.py:448]: done

File: src/sparseml/pytorch/sparsification/pruning/modifier_pruning_rigl.py [hunk 5]
```diff
@@ -0,0 +1,599 @@
...
+from sparseml.utils import FROM_PARAM_TOKEN
+
+
+__all__ = ["RigLPruningModifier", "RigLPruningParamsScorer"]
+
+
+_LOGGER = logging.getLogger(__name__)
+
+
+def cosine_schedule(t: float, t_max: float, init_value: float, end_value: float):
```
[the remark this answers is not in the document]
Comment (ohaijen) [reply src/sparseml/pytorch/sparsification/pruning/modifier_pruning_rigl.py:50]: I am not sure what this comment means. The cosine schedule is not used to set the sparsity (that's constant), but rather to set the proportion of the mask that should be updated, so it doesn't really relate to the get_applied_sparsity_for_epoch method.

File: src/sparseml/pytorch/sparsification/pruning/modifier_pruning_rigl.py (hunk 1 above)
Comment (ohaijen) [reply src/sparseml/pytorch/sparsification/pruning/modifier_pruning_rigl.py:441]: If I understand correctly, this method only gets called when masks should be updated, (when update_ready is true). So we should not be sampling or accumulating any gradients during those steps when the masks are not updated.

File: src/sparseml/pytorch/sparsification/pruning/modifier_pruning_rigl.py [hunk 6]
```diff
@@ -0,0 +1,599 @@
...
+            scores = [
+                self.get_param_score(param, param_grad, param_sparsity)
+                for param, param_grad, param_sparsity in zip(
+                    self._params, self._param_grads, self._sparsity_distribution
+                )
+            ]
+
+        # free memory
+        for i, _ in enumerate(self._param_grads):
+            self._param_grads[i] = None
```
[the remark this answers is not in the document]
Comment (ohaijen) [reply src/sparseml/pytorch/sparsification/pruning/modifier_pruning_rigl.py:597]: No, the comment was misleading. This actually just empties the self._param_grads cache. I changed the comment.

File: src/sparseml/pytorch/sparsification/pruning/modifier_pruning_rigl.py [hunk 7]
```diff
@@ -0,0 +1,599 @@
...
+    Pruner supports three sparsity strategies as in the original paper:
+    - uniform
+    - erdos_renyi
+    - erdos_renyi_kernel
+
+    Supported mask types: unstructured.
+
+    | Sampl' yaml:
+    |   !RigLPruningModifier
+    |       sparsity: 0.7
```
[the remark this answers is not in the document]
Comment (ohaijen) [reply src/sparseml/pytorch/sparsification/pruning/modifier_pruning_rigl.py:103]: done

File: src/sparseml/pytorch/torchvision/train.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (anmarques) [reply src/sparseml/pytorch/torchvision/train.py]: You can leave the TODO in place.

File: src/sparseml/pytorch/sparsification/pruning/modifier_pruning_rigl.py [hunk 8]
```diff
@@ -0,0 +1,599 @@
...
+        # We drop and replace update_fraction of nonzero elements
+        updated_number = (1 - param_sparsity) * self._update_fraction
+        magn_score = param.abs()
+        # Of the existing mask, we keep the top 1-param_sparsity-updated_number
+        # elements by magnitude.
+        magn_score = threshold_fraction(magn_score, param_sparsity + updated_number)
+        # We fill in the mask by also adding the updated_number weights
+        # of the ones that are currently masked, using the gradient magnitude
+        # as the criterion..
+        grad_score = param_grad.abs() * magn_score.eq(0)
```
[the remark this answers is not in the document]
Comment (anmarques) [reply src/sparseml/pytorch/sparsification/pruning/modifier_pruning_rigl.py:576]: I mean, don't we have access to the actual variable that stores the mask, instead of looking at the weights that are zero?

File: src/sparseml/pytorch/sparsification/pruning/modifier_pruning_rigl.py [hunk 9]
```diff
@@ -0,0 +1,600 @@
...
+    threshold, _ = torch.kthvalue(tensor.reshape(-1), k=lookup_idx)
+    return torch.where(tensor > threshold, 1.0, 0.0)
+
+
+@PyTorchModifierYAML()
+class RigLPruningModifier(BaseGradualPruningModifier):
+    """
+    As described in https://arxiv.org/abs/1911.11134
+
+    Sparse training procedure that starts from a sparse model
```
[the remark this answers is not in the document]
Comment (ohaijen) [reply src/sparseml/pytorch/sparsification/pruning/modifier_pruning_rigl.py:85]: fixed.

File: src/sparseml/pytorch/sparsification/pruning/modifier_pruning_rigl.py [hunk 10]
```diff
@@ -0,0 +1,600 @@
...
+    |       inter_func: "cosine"
+    |       global_sparsity: True
+    |       mask_type: unstructured
+    |       sparsity_strategy: "erdos_renyi_kernel"
+    |       init_update_fraction: 0.3
+    |       grad_sampler_kwargs:
+    |           batch_size: 256
+
+    :param final_sparsity: the final sparsity for the param to end with at end_epoch.
+        A single float for the whole model. Note that with RigL, init_sparsity is
```
[the remark this answers is not in the document]
Comment (ohaijen) [reply src/sparseml/pytorch/sparsification/pruning/modifier_pruning_rigl.py:119]: fixed

File: src/sparseml/pytorch/sparsification/pruning/modifier_pruning_rigl.py [hunk 11]
```diff
@@ -0,0 +1,600 @@
...
+        )
+
+    @staticmethod
+    def _reset_momentum_buffer(optimizer):
+        # multiply momentum buffer my param mask
+        if "state" in optimizer.state_dict():
+            for param_group in optimizer.param_groups:
+                for param in param_group["params"]:
+                    state = optimizer.state[param]
+                    param_mask = param.ne(0)
```
[the remark this answers is not in the document]
Comment (ohaijen) [reply src/sparseml/pytorch/sparsification/pruning/modifier_pruning_rigl.py:397]: We do. But I don't think we have a straightforward way of matching optimizer params with the pruning masks, so this ultimately seemed simpler. Please correct me if I'm wrong.

[the remark this answers is not in the document]
Comment (anmarques) [reply src/sparseml/pytorch/sparsification/pruning/modifier_pruning_rigl.py:389]: Oh I see, these are the optimizer params.

**Review by anmarques — DISMISSED** [review]

**Review by natuan — DISMISSED** [review]

**Review by anmarques — APPROVED** [review]

**Review by natuan — APPROVED** [review]
